XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,419 @@
+available when the connection's constructor is executed. Meaning there would be
+no `parentId` available on the JS stack. Instead the TCP server's unique id is
+manually passed to the client's constructor and propagated to the `init()`'s
+`parentId`.
+
+
+#### `pre(id, entryPoint)`
+
+* `id` {Number}
+* `entryPoint` {Function}
+
+Called just before the return callback is called after completing an
+asynchronous request. Or called on handles with events such as receiving a new
+connection. For requests, such as `fs.open()`, this should be called exactly
+once. For handles, such as a TCP server, this may be called 0-N times.
+
+The `entryPoint` is the callback that will be executed immediately after
+`pre()` returns. This is being passed as a way to track what work will be done.
+
+**Note(trevnorris):** I'm not sure passing only `entryPoint` is useful. Since
+the arguments of the function to be called will be available it may be useful
+to pass those to `pre()` as well. Though I have reservations about the user
+being able to mess with non-primitives.
+
+
+#### `post(id, didThrow)`
+
+* `id` {Number}
+* `didThrow` {Boolean}
+
+Called immediately after the return callback is completed. If the callback
+threw but was caught by a domain or `uncaughtException`, `didThrow` will be set
+to `true`. If the callback threw but was not caught then the process will exit
+immediately without calling `post()`.
+
+
+#### `destroy(id)`
+
+* `id` {Number}
+
+Called either when the class destructor is run, or if the resource is marked as
+free. The destructor will usually run when explicitly called (the case for
+handles) or when a request has completed. In several cases this will be
+triggered by GC. In the case of shared or cached resources, such as
+`HTTPParser`, `destroy()` will be called manually when the TCP connection is no
+longer in need of it. Every subsequent use of a shared resource will have a new
+unique id.
+
+
+## API Exceptions
+
+### net Client connection Event
+
+Technically the `pre()`/`post()` events of the `'connection'` event for
+`net.Server` would place the server as the active id. Problem is that this is
+not intuitive in how the asynchronous chain would propagate. So instead make
+the client the active id for the duration of the `'connection'` callback.
+
+
+### Reused Resources
+
+Resources like `HTTPParser` are reused throughout the lifetime of the
+application. This means node will have to synthesize the `init()` and
+`destroy()` calls. Also the id on the class instance will need to be changed
+every time the resource is acquired for use.
+
+For shared resources like `TimerWrap` this is not necessary since there is a
+unique JS handle that will contain the unique id necessary for the calls.
+
+
+## Notes
+
+### Promises
+
+Node currently doesn't have sufficient API to notify calls to Promise
+callbacks. In order to do so node would have to override the native
+implementation.
```

Comment:
... "we are working with the V8 team to gain access to enough internals to integrate Promises into the AsyncWrap API at a future date" or something to that effect. Current wording sounds like a brush-off which it's not.